Source/WebCore/rendering/svg/RenderSVGPath.cpp

```diff
@@ -177,7 +177,7 @@ bool RenderSVGPath::isRenderingDisabled() const
 {
     // For a polygon, polyline or path, rendering is disabled if there is no path data.
     // No path data is possible in the case of a missing or empty 'd' or 'points' attribute.
-    return path().isEmpty();
+    return !hasPath() || path().isEmpty();
```

Comment:
I saw the assertion in RenderSVGShape::path() in https://bugs.webkit.org/show_bug.cgi?id=149613. And this assertion could have been fixed by the same fix you did. But it turned out to be because of a serious problem that should be fixed somewhere else.

I thought about adding this check but I was afraid this will hide many bugs that will be hard to be discovered. So maybe looking at the test case you are investigating can reveal a serious bug.

Reply:
It's not a simple test unfortunately, it happens to me when loading element app (matrix client). I'll get a backtrace.

Reply:
I need a bt ;-)

Reply:
I agree with @shallawa we should try hard to understand the exact condition that leads to this.

Reply:
```
#0  0x00007f181edf5b20 in WebCore::Path::isEmpty() const () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#1  0x00007f181f35bbab in WebCore::SVGRenderSupport::computeContainerBoundingBoxes(WebCore::RenderElement const&, WebCore::FloatRect&, bool&, WebCore::FloatRect&, WebCore::FloatRect&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#2  0x00007f181f38197d in WebCore::LegacyRenderSVGRoot::updateCachedBoundaries() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#3  0x00007f181f381eef in WebCore::LegacyRenderSVGRoot::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#4  0x00007f181f14d5aa in WebCore::RenderFlexibleBox::layoutAndPlaceChildren(WebCore::LayoutUnit&, WTF::Vector<WebCore::FlexItem, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit, bool, WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#5  0x00007f181f14f134 in WebCore::RenderFlexibleBox::layoutFlexItems(bool) [clone .part.0] () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#6  0x00007f181f14faa9 in WebCore::RenderFlexibleBox::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#7  0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#8  0x00007f181f14d5aa in WebCore::RenderFlexibleBox::layoutAndPlaceChildren(WebCore::LayoutUnit&, WTF::Vector<WebCore::FlexItem, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit, bool, WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#9  0x00007f181f14f134 in WebCore::RenderFlexibleBox::layoutFlexItems(bool) [clone .part.0] () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#10 0x00007f181f14faa9 in WebCore::RenderFlexibleBox::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#11 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#12 0x00007f181f10c9c0 in WebCore::RenderBlock::layoutPositionedObject(WebCore::RenderBox&, bool, bool) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#13 0x00007f181f0c4bbb in WebCore::RenderBlock::layoutPositionedObjects(bool, bool) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#14 0x00007f181f11ccf7 in WebCore::RenderBlockFlow::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#15 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#16 0x00007f181f10f30a in WebCore::RenderBlockFlow::layoutBlockChild(WebCore::RenderBox&, WebCore::RenderBlockFlow::MarginInfo&, WebCore::LayoutUnit&, WebCore::LayoutUnit&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#17 0x00007f181f11ac2d in WebCore::RenderBlockFlow::layoutBlockChildren(bool, WebCore::LayoutUnit&) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#18 0x00007f181f11c4de in WebCore::RenderBlockFlow::layoutInFlowChildren(bool, WebCore::LayoutUnit&, WebCore::LayoutUnit&, WebCore::LayoutUnit&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#19 0x00007f181f11ca34 in WebCore::RenderBlockFlow::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#20 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#21 0x00007f181f14eb1e in WebCore::RenderFlexibleBox::alignChildren(WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#22 0x00007f181f14ebd7 in WebCore::RenderFlexibleBox::repositionLogicalHeightDependentFlexItems(WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#23 0x00007f181f14f38d in WebCore::RenderFlexibleBox::layoutFlexItems(bool) [clone .part.0] () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#24 0x00007f181f14faa9 in WebCore::RenderFlexibleBox::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#25 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#26 0x00007f181eb97c5e in WebCore::LocalFrameViewLayoutContext::performLayout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#27 0x00007f181eb9815d in WebCore::LocalFrameViewLayoutContext::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#28 0x00007f181e42146e in WebCore::Document::updateLayout(WTF::OptionSet<WebCore::LayoutOptions>, WebCore::Element const*) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#29 0x00007f181e43f5a9 in WebCore::Element::offsetTop() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#30 0x00007f181e455b7e in WebCore::Element::offsetTopForBindings() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#31 0x00007f181d42a24d in WebCore::jsHTMLElement_offsetTop(JSC::JSGlobalObject*, long, JSC::PropertyName) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#32 0x00007f181af9579b in JSC::PropertySlot::customGetter(JSC::VM&, JSC::PropertyName) const () from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#33 0x00007f181ab05ea3 in JSC::LLInt::performLLIntGetByID(JSC::BytecodeIndex, JSC::CodeBlock*, JSC::JSGlobalObject*, JSC::JSValue, JSC::Identifier const&, JSC::GetByIdModeMetadata&) ()
   from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#34 0x00007f181ab06de2 in llint_slow_path_get_by_id () from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#35 0x00007f1819d835b9 in llint_op_get_by_id () from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#36 0x0000000000000000 in ?? ()
```

Reply:
This `m_path` is nullptr in RenderSVGPath. This means that we have never called `updateShapeFromElement` for this class before. In this condition, using `isRenderingDisabled` is wrong since it is relying on the update from `updateShapeFromElement`. Let's imagine SVG style gets updated (like strokeWidth), and `updateShapeFromElement` needs to be called to reflect this to the path. But if we are not calling this, this `m_path` is pointing at the stale, previous `m_path`, or `nullptr`. And this is the bug here.

When looking at upper container's `LegacyRenderSVGRoot::layout`, it is calling `SVGRenderSupport::layoutChildren`. And we are skipping layout when some conditions are met.
This means that, when computing bounding box, we need to check the exact same condition to skip these non-layout renderers as `isRenderingDisabled`. Current `isRenderingDisabled` is relying on `m_path`, which is only effective after the layout. So, `isRenderingDisabled` implementation looks wrong, this should check both,

1. If layout happens, then it should check m_path's emptiness. This is required by SVG's spec IIUC (empty path should not be rendered)
2. If layout is skipped, then it should say false. So we need to check the same condition to `SVGRenderSupport::layoutChildren`'s layout skipping.

Reply:
We now know that some instances of this crash (e.g. on bing.com) are related to `content-visibility: auto`